File: fe/src/lib/services/messaging.ts
import { api, buildApiUrl } from './api';

export interface Channel {
    id: string;
    name: string;
    description?: string;
    is_private: boolean;
    created_at: string;
    created_by?: string;
    dm_other_user_name?: string | null;
}

export interface ChannelMember {
    id: string;
    display_name: string;
    email: string;
}

export interface Attachment {
    id: string;
    message_id: string;
    file_name: string;
    content_type: string;
    file_size: number;
    is_image: boolean;
    created_at: string;
}

export interface Message {
    id: string;
    channel_id: string;
    sender_id?: string;
    sender_name?: string;
    content: string;
    created_at: string;
    attachments: Attachment[];
}

export interface ChannelMediaItem extends Attachment {
    sender_id?: string;
    sender_name?: string;
}

export const messagingService = {
    async listChannels(): Promise<Channel[]> {
        const res = await api.get('/messaging/channels');
        if (!res.ok) throw new Error('Failed to load channels');
        return res.json();
    },

    async getChannel(channelId: string): Promise<Channel> {
        const res = await api.get(`/messaging/channels/${channelId}`);
        if (!res.ok) throw new Error('Failed to get channel');
        return res.json();
    },

    async createChannel(data: { name: string; description?: string; is_private: boolean; members?: string[] }): Promise<Channel> {
        const res = await api.post('/messaging/channels', data);
        if (!res.ok) throw new Error('Failed to create channel');
        return res.json();
    },

    async addMember(channelId: string, userId: string): Promise<void> {
        const res = await api.post(`/messaging/channels/${channelId}/members`, { user_id: userId });
        if (!res.ok) throw new Error('Failed to add member');
    },

    async listMessages(channelId: string): Promise<Message[]> {
        const res = await api.get(`/messaging/channels/${channelId}/messages`);
        if (!res.ok) throw new Error('Failed to load messages');
        return res.json();
    },

    async sendMessage(
        channelId: string,
        content: string,
        files: File[] = [],
        parentId?: string,
    ): Promise<Message> {
        const formData = new FormData();
        formData.set('content', content);
        if (parentId) formData.set('parentId', parentId);
        for (const file of files) {
            formData.append('files', file);
        }
        const res = await api.postForm(`/messaging/channels/${channelId}/messages`, formData);
        if (!res.ok) {
            const text = await res.text();
            throw new Error(text || 'Failed to send message');
        }
        return res.json();
    },

    async listChannelMedia(channelId: string): Promise<ChannelMediaItem[]> {
        const res = await api.get(`/messaging/channels/${channelId}/media`);
        if (!res.ok) throw new Error('Failed to load shared media');
        return res.json();
    },

    attachmentUrl(channelId: string, attachmentId: string): string {
        return buildApiUrl(`/messaging/channels/${channelId}/attachments/${attachmentId}`);
    },

    async fetchAttachmentBlob(channelId: string, attachmentId: string): Promise<string> {
        const token =
            typeof window !== 'undefined' ? localStorage.getItem('access_token') : null;
        const res = await fetch(this.attachmentUrl(channelId, attachmentId), {
            headers: token ? { Authorization: `Bearer ${token}` } : {},
        });
        if (!res.ok) throw new Error('Failed to load attachment');
        const blob = await res.blob();
        return URL.createObjectURL(blob);
    },

    async updateChannel(channelId: string, data: { name?: string; description?: string }): Promise<Channel> {
        const res = await api.put(`/messaging/channels/${channelId}`, data);
        if (!res.ok) throw new Error('Failed to update channel');
        return res.json();
    },

    async getChannelMembers(channelId: string): Promise<ChannelMember[]> {
        const res = await api.get(`/messaging/channels/${channelId}/members`);
        if (!res.ok) throw new Error('Failed to list members');
        return res.json();
    },

    async removeMember(channelId: string, userId: string): Promise<void> {
        const res = await api.delete(`/messaging/channels/${channelId}/members/${userId}`);
        if (!res.ok) throw new Error('Failed to remove member');
    }
};

File: mobile/lib/models/messaging.dart
class Channel {
  Channel({
    required this.id,
    required this.name,
    this.description,
    required this.isPrivate,
    required this.createdAt,
    this.createdBy,
    this.dmOtherUserName,
  });

  final String id;
  final String name;
  final String? description;
  final bool isPrivate;
  final DateTime createdAt;
  final String? createdBy;
  final String? dmOtherUserName;

  /// A 1:1 DM channel is created private with exactly the two members; the
  /// backend resolves the other participant's name into `dm_other_user_name`
  /// per-viewer. Group channels and public channels never have it set.
  bool get isDirectMessage => isPrivate && dmOtherUserName != null;

  String displayName() => isDirectMessage ? dmOtherUserName! : name;

  factory Channel.fromJson(Map<String, dynamic> json) => Channel(
        id: json['id'] as String,
        name: json['name'] as String,
        description: json['description'] as String?,
        isPrivate: json['is_private'] as bool? ?? false,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
        createdBy: json['created_by'] as String?,
        dmOtherUserName: json['dm_other_user_name'] as String?,
      );
}

class Attachment {
  Attachment({
    required this.id,
    required this.messageId,
    required this.fileName,
    required this.contentType,
    required this.fileSize,
    required this.isImage,
    required this.createdAt,
  });

  final String id;
  final String messageId;
  final String fileName;
  final String contentType;
  final int fileSize;
  final bool isImage;
  final DateTime createdAt;

  factory Attachment.fromJson(Map<String, dynamic> json) => Attachment(
        id: json['id'] as String,
        messageId: json['message_id'] as String,
        fileName: json['file_name'] as String,
        contentType: json['content_type'] as String,
        fileSize: json['file_size'] as int,
        isImage: json['is_image'] as bool? ?? false,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
      );
}

class ReactionSummary {
  ReactionSummary({required this.emoji, required this.count, required this.reactedByMe});

  final String emoji;
  final int count;
  final bool reactedByMe;

  factory ReactionSummary.fromJson(Map<String, dynamic> json) => ReactionSummary(
        emoji: json['emoji'] as String,
        count: json['count'] as int,
        reactedByMe: json['reacted_by_me'] as bool? ?? false,
      );
}

class Message {
  Message({
    required this.id,
    required this.channelId,
    this.senderId,
    this.senderName,
    required this.content,
    required this.createdAt,
    this.attachments = const [],
    this.reactions = const [],
  });

  final String id;
  final String channelId;
  final String? senderId;
  final String? senderName;
  final String content;
  final DateTime createdAt;
  final List<Attachment> attachments;
  final List<ReactionSummary> reactions;

  Message copyWithReactions(List<ReactionSummary> newReactions) => Message(
        id: id,
        channelId: channelId,
        senderId: senderId,
        senderName: senderName,
        content: content,
        createdAt: createdAt,
        attachments: attachments,
        reactions: newReactions,
      );

  factory Message.fromJson(Map<String, dynamic> json) => Message(
        id: json['id'] as String,
        channelId: json['channel_id'] as String,
        senderId: json['sender_id'] as String?,
        senderName: json['sender_name'] as String?,
        content: json['content'] as String,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
        attachments: (json['attachments'] as List<dynamic>? ?? [])
            .map((e) => Attachment.fromJson(e as Map<String, dynamic>))
            .toList(),
        reactions: (json['reactions'] as List<dynamic>? ?? [])
            .map((e) => ReactionSummary.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class ChannelMember {
  ChannelMember({required this.id, required this.displayName, required this.email});

  final String id;
  final String displayName;
  final String email;

  factory ChannelMember.fromJson(Map<String, dynamic> json) => ChannelMember(
        id: json['id'] as String,
        displayName: json['display_name'] as String,
        email: json['email'] as String,
      );
}

class ChannelMediaItem {
  ChannelMediaItem({
    required this.id,
    required this.messageId,
    required this.fileName,
    required this.contentType,
    required this.fileSize,
    required this.isImage,
    required this.createdAt,
    this.senderId,
    this.senderName,
  });

  final String id;
  final String messageId;
  final String fileName;
  final String contentType;
  final int fileSize;
  final bool isImage;
  final DateTime createdAt;
  final String? senderId;
  final String? senderName;

  factory ChannelMediaItem.fromJson(Map<String, dynamic> json) => ChannelMediaItem(
        id: json['id'] as String,
        messageId: json['message_id'] as String,
        fileName: json['file_name'] as String,
        contentType: json['content_type'] as String,
        fileSize: json['file_size'] as int,
        isImage: json['is_image'] as bool? ?? false,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
        senderId: json['sender_id'] as String?,
        senderName: json['sender_name'] as String?,
      );
}
